test/com/cdd/bao/validator/RemappingCheckerTest.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * BioAssay Ontology Annotator Tools
+ * 
+ * (c) 2017-2018 Collaborative Drug Discovery Inc.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License 2.0
+ * as published by the Free Software Foundation:
+ * 
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html
+ * 
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+package com.cdd.bao.validator;
+
+import static org.junit.Assert.*;
+
+import java.io.*;
+import java.util.*;
+
+import org.junit.*;
+
+/*
+	Test for logic in RemappingChecker that vets remapping chains
+*/
+
+public class RemappingCheckerTest
+{
+	@Test
+	public void testValidOneLinkChain()
+	{
+		// a => b (valid)
+		Map<String, String> chain = new HashMap<>();
+		chain.put("a", "b");
+		try { RemappingChecker.validateRemappings(chain); }
+		catch (IOException ioe) { assertTrue("No exception should occur here!", false); }
+		assertTrue("Invalid chain leads to cycle.", true);
+	}
+
+	@Test
+	public void testValidTwoLinkChain()
+	{
+		// a => b => c (valid)
+		Map<String, String> chain = new HashMap<>();
+		chain.put("a", "b");
+		chain.put("b", "c");
+		try { RemappingChecker.validateRemappings(chain); }
+		catch (IOException ioe) { assertTrue("No exception should occur here!", false); }
+		assertTrue("Invalid chain leads to cycle.", true);
+	}
```

Comment:
There is a lot of code repetition here and with all the exception handling it's hard to understand if the test is for a valid or an invalid case. You could move the actual testing into a separate method:
```
private void assertValidRemapping(Map<String, String> chain, boolean expectFailure)
{
  try
  {
    RemappingChecker.validateRemappings(chain);
    if (expectFailure) throw new AssertionError("Invalid chain passes: " + chain);
  }
  catch (IOException e)
  {
    if (!expectFailure) throw new AssertionError("Valid chain fails: " + chain, e)
  }
}
```
Adding the `chain` map to the message gives better feedback if something goes wrong.

I would also add a method that takes a string with the chain definition and returns the mapping
```
private getChainMap(String mapping)
{
  Map<String, String> chain = new HashMap<>();
  for (String s : mapping.split(",")) chain.put(s.split("=>")[0], s.split("=>")[1];
  return chain;
}
```

This would give the following code:
```
@Test
public void testChainValidation()
{
  assertValidRemapping(getChainMap("a=>b", false);
  assertValidRemapping(getChainMap("a=>a", true);
  assertValidRemapping(getChainMap("a=>b,b=>c", false);
  assertValidRemapping(getChainMap("a=>b,b=>a", true);
  // two chains with one invalid
  assertValidRemapping(getChainMap("a=>b,b=>c,d=>e,e=>d", true);
  // a => b => c => d => e => c (invalid)
  assertValidRemapping(getChainMap("a=>b,b=>c,c=>d,d=>e,e=>c", true);

  Map<String, String> chain = new HashMap<>();
  chain.put("a", null);
  assertValidRemapping(chain, true); // don't know what to expect here
}
```

Reply:
I'm open to a wider discussion on how best to author unit-tests.  I would prefer, however, to leave things as-is for now.